### PR TITLE
server,settings: properly cascade defaults for TenantReadOnly

### LINKED ISSUE #85729: settings: redefine TenantReadOnly to take its value from the system tenant
As discussed with @stevendanna @dt and @ajwerner : 

-  we want a setting class which uses a single shared value between system and secondary tenants.
- the current behavior of TenantReadOnly (one value per tenant) can be readily emulated by a TenantReadWrite setting + an OVERRIDE in the system tenant.

So to achieve point 1, we can redefine TenantReadOnly to make settings take their value from the system tenant.

Informs #77472

Epic: CRDB-6671

Jira issue: CRDB-18406

### LINKED ISSUE #91825: *: reads of SystemOnly settings from tenant process
**Describe the problem**

SystemOnly settings are not intended to be read from a tenant process.  However, nothing actually disallows a tenant process from reading the in-memory default of a SystemOnly setting and taking actions on it.

The settings implementation includes a safety check for this was intended to be enabled in test builds, but that safety check requires `(*Values).SetNonSystemTenant` is called at some point during tenant startup.  Currently we do not call that method so no accesses to SystemOnly settings from tenants are disallowed.

This recently tripped us up here https://github.com/cockroachdb/cockroach/pull/91824 and a quick test run with SetNonSystemTenant immediately hit our safety check on `scheduler_latency.sample_period`.  It is likely that there are a number of other SystemOnly settings being accessed from tenant processes.

We should enable SetNonSystemTenant in our tests and work through the problems.

Jira issue: CRDB-21443
Epic: CRDB-6671

### LINKED ISSUE #108677: settings: make host cluster or KV settings accessible to tenants
We have some settings that it only makes sense to set for the host cluster, but that SQL code still needs to be able to read. There's currently no way to do this, and I see indications that a lot of code thinks it does but actually doesn't. We have three settings:

* `SystemOnly`: can only be set and read by the host cluster.
* `TenantReadOnly`: tenants can read the setting, but the setting must be set individually for each tenant. The host cluster setting is invisible, the tenant will see the default setting unless an individual setting is set.
* `TenantWritable`: the setting is tenant-specific.

Consider e.g. `kv.closed_timestamp.side_transport_interval` which controls how often we send closed timestamp updates from leaseholders to followers. This is currently `TenantWritable`, which doesn't make any sense because the side transport doesn't run in a tenant, it runs below KV. However, SQL code (in this case changefeeds) still needs to access this setting, for example because it [needs to know how often to expect closed timestamps to be emitted](https://github.com/cockroachdb/cockroach/blob/72a8b4e40b107b2de8cd402983f85e62c42b4753/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go#L677-L687). Currently, they won't see the KV setting, they'll see the tenant setting, which can lead to changefeeds breaking completely in tenants with certain configurations. But if I change it to `SystemOnly` the setting isn't visible at all (well, code can read it because of a [bug](https://github.com/cockroachdb/cockroach/issues/91825), but they'll only see the default value not the host value), and if I change it to `TenantReadOnly` it will only read the tenant's version of the setting, not the host cluster.

This seems pretty broken, and it isn't clear to me how to proceed here. Do we need a new class, e.g. `TenantVisible`, or alternatively make `SystemOnly` visible to tenants (at least from code)?

See also [Slack thread](https://cockroachlabs.slack.com/archives/C02HWA24541/p1691924915377419).

This has tripped us up in #91824 and #108678, and also applies to e.g. `kv.rangefeed.enabled`, `follower_read_timestamp()`, and many other settings.

Jira issue: CRDB-30584
Epic: CRDB-6671

### PR DESCRIPTION
**Pull request by knz**
Previous in sequence:
- [x] #110789 and #110947 
- [x] #110676
- [x] #111008
- [x] #111145
- [x] #111147
- [x]  #111149
- [x] #111150
- [x] #111153
- [x] #111475
- [x] #111210
- [x] #111212

Fixes #108677.
Fixes #85729.
Fixes #91825.
Completes the work described in the settings RFC.
Epic: CRDB-6671

TLDR: this patch ensures that virtual cluster servers observe changes
made to TenantReadOnly settings via SET CLUSTER SETTING in the system
interface, even when there is no override set via ALTER VIRTUAL
CLUSTER SET CLUSTER SETTING.

For example, after `SET CLUSTER SETTING
kv.closed_timestamp.target_duration = '10s'` in the system interface,
this value will show up via `SHOW CLUSTER SETTING` in a virtual
cluster SQL session.

This changes the way that settings are picked up in virtual cluster,
as follows:

1. if there is an override specifically for this tenant's ID
   (in `tenant_settings`), use that.
2. otherwise, if there is an override for the pseudo-ID 0
   (in `tenant_settings` still, set via `ALTER TENANT ALL SET CLUSTER
   SETTING`), then use that.
3. **NEW**  otherwise, if the class is TenantReadOnly and there
   is a custom value in `system.settings`, set via a regular
   `SET CLUSTER SETTING` in the system tenant, then use that.
4. otherwise, use the global default set via the setting's
   `Register()` call.

----

Prior to this patch, TenantReadOnly settings as observed from virtual
clusters were defined as the following priority order:

1. if there is an override specifically for this tenant's ID
   (in `tenant_settings`), use that.
2. otherwise, if there is an override for the pseudo-ID 0
   (in `tenant_settings` still, set via `ALTER TENANT ALL SET CLUSTER
   SETTING`), then use that.
3. otherwise, use the global default set via the setting's
   `Register()` call.

Remarkably, this did not pick up any changes made via a plain `SET
CLUSTER SETTING` statement via the system interface, which only
modifies this setting's value in `system.settings` (thus not
`tenant_settings`).

This situation was problematic in two ways.

To start, settings like `kv.closed_timestamp.target_duration` cannot
be set solely in `system.tenant_settings`; they are also used in the
storage layer and so must also be picked up from changes in
`system.settings`.

For these settings, it is common for operators to just issue the plain
`SET CLUSTER SETTING` statement (to update `system.settings`) and
simply forget to _also_ run `ALTER TENANT ALL SET CLUSTER SETTING`.

This mistake is nearly unavoidable and would result in incoherent
behavior, where the storage layer would use the customized value
and virtual clusters would use the registered global default.

The second problem is in mixed-version configurations, where
the storage layer runs version N+1 and the SQL service runs version N
of the executable. If the registered global default changes from
version N to N+1, the SQL service would not properly pick up
the new default defined in version N+1 of the storage layer.

This patch fixes both problems as follows:

- it integrates changes to TenantReadOnly settings observed
  in `system.settings`, to the watcher that tracks changes
  to `system.tenant_settings`. When a TenantReadOnly setting
  is present in the former but not the latter, a synthetic
  override is added.

- it also initializes synthetic overrides for all the
  TenantReadOnly settings upon server initialization,
  from the registered global default, so that virtual
  cluster servers always pick up the storage layer's
  default as override.

### CODE REVIEW

**Comment by knz** [comment]
Next steps:
- add missing unit tests
- remove now-redundant calls to .Override in various tests

**Comment by knz** [comment]
let's keep it as draft for now.

**Review by yuzefovich — COMMENTED** [review]
Nice work figuring all of this out! I skipped the first commit, but 2nd through 4th look good to me, I only have a handful of comments. It seems beneficial for Steven to review too.

> Reviewed 72 of 73 files at r9, 3 of 3 files at r10, 2 of 2 files at r11, 80 of 80 files at r12, 3 of 3 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-:-NfJDjbv5hjGL5dZjoCT:bege24f)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @stevendanna)

___
*[`pkg/server/tenantsettingswatcher/overrides_store.go` line 309 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfIp4-WEB3_moG8qYDg:-NfIp4-WEB3_moG8qYDh:b-3ol4n7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7373857b2bb22838f5391bab3aace81d06d8d780/pkg/server/tenantsettingswatcher/overrides_store.go#L309)):*
> ```Go
> 	alternateDefaults []kvpb.TenantSetting,
> ) []kvpb.TenantSetting {
> 	// log.VEventf(context.TODO(), 1, "splicing alternate defaults:\nexplicitKeys=%# v\noverrides=%# v\nalternateDefaults=%# v",
> ```

nit: leftover here and throughout this function.
___
*[`pkg/server/tenantsettingswatcher/overrides_store.go` line 315 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfIpkd9-B4SBnDiT5qv:-NfIpkd9-B4SBnDiT5qw:bb744yp) ([raw file](https://github.com/cockroachdb/cockroach/blob/7373857b2bb22838f5391bab3aace81d06d8d780/pkg/server/tenantsettingswatcher/overrides_store.go#L315)):*
> ```Go
> 	aIter, bIter := 0, 0
> 	for aIter < len(overrides) && bIter < len(alternateDefaults) {
> 		log.VEventf(context.Background(), 1, "iter: aIter=%d bIter=%d", aIter, bIter)
> ```

nit: we should have `ctx` available in scope of all callsites, but also I don't think you want to check this in, right?
___
*[`pkg/server/tenantsettingswatcher/overrides_store.go` line 385 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfIr7LBA60Q5W62kJU4:-NfIr7LBA60Q5W62kJU5:blaza2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/7373857b2bb22838f5391bab3aace81d06d8d780/pkg/server/tenantsettingswatcher/overrides_store.go#L385)):*
> ```Go
> 		} else {
> 			numDst := 0
> 			for ; (aIter+numDst) < len(dst) &&
> ```

I'm confused here about the conditions. Why are we checking `aIter+numDst < len(dst)`? Also, why are we looking at `dst[aIter+numDst]` instead of `dst[aiter+numDst` in the last condition?

We're in the case when `dst[aIter].InternalKey > src[bIter].InternalKey` which means that we have some elements in `src` that are not present in `dst`. My expectation is that we would simply copy those elements over by inserting them into `dst` before `aIter`. This loop is about counting the number of contiguous elements in `src`, so I would expect the code to look roughly like a similar pattern in `spliceOverrideDefaults`. What is different here?
___
*[`pkg/server/tenantsettingswatcher/overrides_store_test.go` line 86 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfIsHQYBPPKgZIMnIPz:-NfIsHQYBPPKgZIMnIQ-:bvasy9k) ([raw file](https://github.com/cockroachdb/cockroach/blob/7373857b2bb22838f5391bab3aace81d06d8d780/pkg/server/tenantsettingswatcher/overrides_store_test.go#L86)):*
> ```Go
> 	expect(o3, "x=xx")
> 
> 	// Verify hat overrides also work for the special "all tenants" ID.
> ```

nit: `s/ hat / that /`.
___
*[`pkg/server/tenantsettingswatcher/overrides_store_test.go` line 153 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfIsASZ6vEiOlzNrsgm:-NfIsASZ6vEiOlzNrsgn:bucimge) ([raw file](https://github.com/cockroachdb/cockroach/blob/7373857b2bb22838f5391bab3aace81d06d8d780/pkg/server/tenantsettingswatcher/overrides_store_test.go#L153)):*
> ```Go
> }
> 
> func TestSpliceOverrideDefaults(t *testing.T) {
> ```

I think we also need a unit test for `updateDefaults` since it too has non-trivial logic.
___
*[`pkg/settings/integration_tests/read_only_1/propagation_test.go` line 23 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfJDKLM3UXdOqZgSwYi:-NfJDKLM3UXdOqZgSwYj:b-7rofm8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f7e41f9c035e327a944a7583e267ccd42e07655/pkg/settings/integration_tests/read_only_1/propagation_test.go#L23)):*
> ```Go
> var roS = settings.RegisterStringSetting(settings.TenantReadOnly, "tenant.read.only", "desc", "initial")
> 
> func TestSettingDefaultPropagationReadOnly1(t *testing.T) {
> ```

nit: what's the rationale for splitting up each test case into a separate package? I imagine we could have had a single package with multiple separate settings registered. Is it to get more parallelism for speed? Couldn't that be accomplished by having multiple unit tests within a single package?

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110758)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @stevendanna and @yuzefovich)

___
*[`pkg/server/tenantsettingswatcher/overrides_store.go` line 309 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfIp4-WEB3_moG8qYDg:-NfLw5SmDXK6Nobpeyd3:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/7373857b2bb22838f5391bab3aace81d06d8d780/pkg/server/tenantsettingswatcher/overrides_store.go#L309)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: leftover here and throughout this function.
</blockquote></details>

Removed.
___
*[`pkg/server/tenantsettingswatcher/overrides_store.go` line 315 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfIpkd9-B4SBnDiT5qv:-NfLw7St3HVNDugWXjol:bkv2rre) ([raw file](https://github.com/cockroachdb/cockroach/blob/7373857b2bb22838f5391bab3aace81d06d8d780/pkg/server/tenantsettingswatcher/overrides_store.go#L315)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we should have `ctx` available in scope of all callsites, but also I don't think you want to check this in, right?
</blockquote></details>

Indeed. Removed.
___
*[`pkg/server/tenantsettingswatcher/overrides_store.go` line 385 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfIr7LBA60Q5W62kJU4:-NfLwG_R3Vsu-DQ3UV7d:bk89joy) ([raw file](https://github.com/cockroachdb/cockroach/blob/7373857b2bb22838f5391bab3aace81d06d8d780/pkg/server/tenantsettingswatcher/overrides_store.go#L385)):*
> Why are we checking `aIter+numDst < len(dst)`? Also, why are we looking at `dst[aIter+numDst]` instead of `dst[aiter+numDst` in the last condition?

Good catch. This was also a bug in the splice function, which I needed to fix earlier; then I forgot to also remove it here.

This needed a unit test. :crying_cat_face: 
Added the test too.
___
*[`pkg/server/tenantsettingswatcher/overrides_store_test.go` line 86 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfIsHQYBPPKgZIMnIPz:-NfLyfdb1kX-Az-aX1UO:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/7373857b2bb22838f5391bab3aace81d06d8d780/pkg/server/tenantsettingswatcher/overrides_store_test.go#L86)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/ hat / that /`.
</blockquote></details>

Done
___
*[`pkg/server/tenantsettingswatcher/overrides_store_test.go` line 153 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfIsASZ6vEiOlzNrsgm:-NfLygmz4CBNPcf-kZ20:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7373857b2bb22838f5391bab3aace81d06d8d780/pkg/server/tenantsettingswatcher/overrides_store_test.go#L153)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think we also need a unit test for `updateDefaults` since it too has non-trivial logic.
</blockquote></details>

Done.
___
*[`pkg/settings/integration_tests/read_only_1/propagation_test.go` line 23 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfJDKLM3UXdOqZgSwYi:-NfLyifX2jZooEdtad9U:bepclis) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f7e41f9c035e327a944a7583e267ccd42e07655/pkg/settings/integration_tests/read_only_1/propagation_test.go#L23)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: what's the rationale for splitting up each test case into a separate package? I imagine we could have had a single package with multiple separate settings registered. Is it to get more parallelism for speed? Couldn't that be accomplished by having multiple unit tests within a single package?
</blockquote></details>

I spent two days trying to coerce CI to accept the long execution time for the full test (~10mn in CI). I failed; then I gave up. Splitting the test in 4 packages was the only way I found it to pass. (Even just 2 packages was still not enough.)

**Review by yuzefovich — APPROVED** [review]
I only have some minor nits, :lgtm: (don't forget to remove the last commit) I'll leave it up to you to decide whether to wait for Steven's review or not.

> Reviewed 27 of 27 files at r14, 4 of 4 files at r15, 22 of 22 files at r16, 4 of 4 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-:-NfNVARuD61FAsyMTvSy:b-jghz2d)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz and @stevendanna)

___
*[`pkg/server/settingswatcher/settings_watcher.go` line 116 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfNM0Br0S9uwMGfQ5mZ:-NfNM0Br0S9uwMGfQ5m_:b6gmho9) ([raw file](https://github.com/cockroachdb/cockroach/blob/16ca4ff00c501e068c400b4a9e080238107a27ea/pkg/server/settingswatcher/settings_watcher.go#L116)):*
> ```Go
> // NewWithNotifier constructs a new SettingsWatcher which notifies
> // an observer about changes to TenantReadOnly settings.
> func NewWithNotifier(
> ```

nit: perhaps rename it to `NewWithRONotifier`?
___
*[`pkg/server/tenantsettingswatcher/overrides_store.go` line 73 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfNQG399v4WSVgSDP5B:-NfNQG399v4WSVgSDP5C:b7sf8uv) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc617add6caba35d188aa1d5b87635145020d2/pkg/server/tenantsettingswatcher/overrides_store.go#L73)):*
> ```Go
> 	// This map is used when an alternate default is changed
> 	// asynchronously after the overrides have been loaded from the
> 	// rangefeed already, to detect which override need to be updated
> ```

nit: `s/which override/which overrides/`.
___
*[`pkg/server/tenantsettingswatcher/overrides_store.go` line 186 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfNQSxWBT9GaBwAyyu5:-NfNQSxWBT9GaBwAyyu6:bvn5q91) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc617add6caba35d188aa1d5b87635145020d2/pkg/server/tenantsettingswatcher/overrides_store.go#L186)):*
> ```Go
> 	if tenantID == allTenantOverridesID {
> 		// Inherit alternate defaults.
> 		overrides = append([]kvpb.TenantSetting(nil), s.mu.alternateDefaults...)
> ```

nit: not sure how important performance here is, but I think `make + copy` is a little faster that `append`.
___
*[`pkg/server/tenantsettingswatcher/overrides_store.go` line 251 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfNRnQs0giS4pY4rd9I:-NfNRB11ALU-Rrsx5OLp:b-4xv4pd) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc617add6caba35d188aa1d5b87635145020d2/pkg/server/tenantsettingswatcher/overrides_store.go#L251)):*
> ```Go
> // override in system.tenant_settings.
> //
> // The input slice must be sorted by key already.
> ```

nit: do we want to assert this behind `buildutil.CrdbTestBuild` somewhere?
___
*[`pkg/settings/integration_tests/propagation.go` line 37 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfNUaXe0WMUJF3SfX6w:-NfNUaXe0WMUJF3SfX6x:bgy7ya1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b4267173ed2e6a4a4285025df5ed3ae79c5f1/pkg/settings/integration_tests/propagation.go#L37)):*
> ```Go
> 	// This test currently takes >1 minute.
> 	//
> 	// TODO(multi-tenant): make it faster. Currently most of the
> ```

We have 32 configs in which we start a server and a tenant - can we not easily reuse a single server but starting a fresh tenant for each test case?
___
*[`pkg/settings/integration_tests/propagation.go` line 54 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfNUKIg3_oyjyPJbMMe:-NfNUKIg3_oyjyPJbMMf:b3ouvxf) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b4267173ed2e6a4a4285025df5ed3ae79c5f1/pkg/settings/integration_tests/propagation.go#L54)):*
> ```Go
> 	// Speed up the tests.
> 	sysDB.Exec(t, "SET CLUSTER SETTING kv.closed_timestamp.target_duration = '10ms'")
> 	sysDB.Exec(t, "ALTER TENANT ALL SET CLUSTER SETTING kv.closed_timestamp.target_duration = '10ms'")
> ```

nit: isn't this line redundant because the setting is tenant RO?
___
*[`pkg/settings/integration_tests/read_only_1/propagation_test.go` line 23 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfJDKLM3UXdOqZgSwYi:-NfNJ0cM5ydJKhZplGMP:bt8kz8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f7e41f9c035e327a944a7583e267ccd42e07655/pkg/settings/integration_tests/read_only_1/propagation_test.go#L23)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I spent two days trying to coerce bors+CI to accept the long execution time for the full test (~10mn in CI). I failed; then I gave up. Splitting the test in 4 packages was the only way I found it to pass. (Even just 2 packages was still not enough.)
</blockquote></details>

Ack, thanks for the context. (Also, have you tried manually increasing `shard_count` to 4 within a single package? Sorry if this is obvious.)

**Comment by knz** [comment]
>  have you tried manually increasing shard_count to 4 within a single package?

No i have not. I only learned of `shard_count` today. I'll try this next. Not obvious.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110758)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz and @stevendanna)

___
*[`pkg/server/testserver.go` line 1545 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfNYSjf0e5fa29ZnW9S:-NfNYSjf0e5fa29ZnW9T:b-7zds8a) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f60eecd5aaebf176d13eb3e0fe5a47623cd623a/pkg/server/testserver.go#L1545)):*
> ```Go
> 
> 	for _, setting := range []settings.Setting{
> 		sql.SecondaryTenantScatterEnabled,
> ```

nit: it looks like we could unexport these cluster settings (except for MR one) if we replace this loop to use cluster setting names directly since it's the only callsite outside of `sql` that access the objects directly.
___
*[`pkg/settings/integration_tests/propagation.go` line 37 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfNUaXe0WMUJF3SfX6w:-NfNYnlVDl2llef9QmAO:bqz7099) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b4267173ed2e6a4a4285025df5ed3ae79c5f1/pkg/settings/integration_tests/propagation.go#L37)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

We have 32 configs in which we start a server and a tenant - can we not easily reuse a single server but starting a fresh tenant for each test case?
</blockquote></details>

Never mind, we already do that.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110758)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @stevendanna and @yuzefovich)

___
*[`pkg/settings/integration_tests/propagation.go` line 54 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfNUKIg3_oyjyPJbMMe:-NfNX0UCAxcqYsXWcz6O:bnrb5s7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c0b4267173ed2e6a4a4285025df5ed3ae79c5f1/pkg/settings/integration_tests/propagation.go#L54)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: isn't this line redundant because the setting is tenant RO?
</blockquote></details>

lol, yes. :facepalm: 
I blindly copied this from another test. I'll simplify the other tests too:  #111383.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110758)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @stevendanna and @yuzefovich)

___
*[`pkg/server/testserver.go` line 1545 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfNYSjf0e5fa29ZnW9S:-NfWb3rNEW4Q1VPa5bWG:bl7zjo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f60eecd5aaebf176d13eb3e0fe5a47623cd623a/pkg/server/testserver.go#L1545)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: it looks like we could unexport these cluster settings (except for MR one) if we replace this loop to use cluster setting names directly since it's the only callsite outside of `sql` that access the objects directly.
</blockquote></details>

I'm not keen on this; when we do this we lose the ability to do "find references" from the setting definition in a code editor.
___
*[`pkg/server/settingswatcher/settings_watcher.go` line 116 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfNM0Br0S9uwMGfQ5mZ:-NfWaBuUEYtK2pCFJ_IJ:blw9d1r) ([raw file](https://github.com/cockroachdb/cockroach/blob/16ca4ff00c501e068c400b4a9e080238107a27ea/pkg/server/settingswatcher/settings_watcher.go#L116)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: perhaps rename it to `NewWithRONotifier`?
</blockquote></details>

Given it's only called in one place and the comment explains, I'll go for the shorter (and more memorable) name for now.
___
*[`pkg/server/tenantsettingswatcher/overrides_store.go` line 73 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfNQG399v4WSVgSDP5B:-NfQwk-KCGCGVNg9hb5v:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc617add6caba35d188aa1d5b87635145020d2/pkg/server/tenantsettingswatcher/overrides_store.go#L73)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/which override/which overrides/`.
</blockquote></details>

Done.
___
*[`pkg/server/tenantsettingswatcher/overrides_store.go` line 186 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfNQSxWBT9GaBwAyyu5:-NfQwlJm16uPZGMsAFM0:b-gi9xv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc617add6caba35d188aa1d5b87635145020d2/pkg/server/tenantsettingswatcher/overrides_store.go#L186)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: not sure how important performance here is, but I think `make + copy` is a little faster that `append`.
</blockquote></details>

TIL - Done.
___
*[`pkg/server/tenantsettingswatcher/overrides_store.go` line 251 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfNRnQs0giS4pY4rd9I:-NfQxjx2B6t1Q_d5Gu_U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc617add6caba35d188aa1d5b87635145020d2/pkg/server/tenantsettingswatcher/overrides_store.go#L251)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: do we want to assert this behind `buildutil.CrdbTestBuild` somewhere?
</blockquote></details>

Done.

**Comment by knz** [comment]
i've modified the test to use just 1 package with sharding.

**Review by stevendanna — APPROVED** [review]
I'm happy with Yahor's review.

One overall question I found myself asking is whether some of this code would be simpler if the alternate defaults slice was a map. But you have unit test coverage of the relevant functions.

**Comment by knz** [comment]
> One overall question I found myself asking is whether some of this code would be simpler if the alternate defaults slice was a map.

In fact I tried it both ways. It's not so clear -- we do need to keep track of which keys in the map were observed in the existing overrides, so as to iterate over the "remaining keys" to add them if they were not seen yet.

That means 2 maps with a weird iteration at the end. The complexity is still there, just in a different shape.

**Comment by stevendanna** [comment]
@knz Awesome, should have known you already considered it.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110758)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz)

___
*[`pkg/server/testserver.go` line 1545 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfNYSjf0e5fa29ZnW9S:-NfWn2icAcspQf5ZaLQ0:b-7cirll) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f60eecd5aaebf176d13eb3e0fe5a47623cd623a/pkg/server/testserver.go#L1545)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I'm not keen on this; when we do this we lose the ability to do "find references" from the setting definition in a code editor.
</blockquote></details>

Can you expand on "find references" a bit? Like, in GoLand it doesn't matter if a variable is exported or not - I could still search for its uses. (I'm not pushing on this change, trying to understand your rationale.)

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110758)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)

___
*[`pkg/server/testserver.go` line 1545 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/110758#-NfNYSjf0e5fa29ZnW9S:-NfX4m73DkrfISYCa-gq:b-o5a4jj) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f60eecd5aaebf176d13eb3e0fe5a47623cd623a/pkg/server/testserver.go#L1545)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Can you expand on "find references" a bit? Like, in GoLand it doesn't matter if a variable is exported or not - I could still search for its uses. (I'm not pushing on this change, trying to understand your rationale.)
</blockquote></details>

Do you mean that it's always possible to "search for a text string in an entire project" (i.e. `grep`)
Yes that works but it's not exactly precise. And also the lack of structural reference doesn't work very well when we choose to rename settings.

**Comment by knz** [comment]
bors r=stevendanna,yuzefovich

**Comment by yuzefovich** [comment]
> Do you mean that it's always possible to "search for a text string in an entire project" (i.e. grep)
Yes that works but it's not exactly precise.

I meant "Find Usages" feature of GoLand. Not renaming sgtm.
